src/spock_recovery.c

```diff
@@ -442,3 +449,329 @@ update_recovery_slot_progress(const char *slot_name, XLogRecPtr lsn,
+ *
+ * Find the best surviving node to use as a rescue source for a failed origin node.
+ * This function queries all co-subscriber nodes to determine which one has the most
+ * recent transactions from the failed origin node.
+ *
+ * Usage: SELECT * FROM spock.find_rescue_source('failed_node_name');
+ *
+ * Returns a record with:
+ * - origin_node_id: The failed node ID
+ * - source_node_id: The surviving node ID with the most recent data
+ * - last_lsn: The LSN of the last known transaction from the origin
+ * - last_commit_timestamp: The commit timestamp of the last known transaction
+ * - confidence_level: How confident we are in this choice (HIGH/MEDIUM/LOW)
+ */
+Datum
+spock_find_rescue_source_sql(PG_FUNCTION_ARGS)
+{
+	text	   *failed_node_name_text = PG_GETARG_TEXT_PP(0);
+	char	   *failed_node_name;
+	TupleDesc	tupdesc;
+	Datum		values[5];
+	bool		nulls[5];
+	HeapTuple	tuple;
+	Oid			failed_node_id = InvalidOid;
+	Oid			best_source_node_id = InvalidOid;
+	XLogRecPtr	best_lsn = InvalidXLogRecPtr;
+	TimestampTz	best_commit_ts = 0;
+	char	   *confidence_level = "LOW";
+
+	/* Get the failed node name */
+	failed_node_name = text_to_cstring(failed_node_name_text);
+
+	/* Get node ID for the failed node */
+	{
+		SpockNode *node = get_node_by_name(failed_node_name, true);
+		if (node == NULL)
+		{
+			/* Return empty result for non-existent nodes */
+			pfree(failed_node_name);
+			PG_RETURN_NULL();
+		}
+		failed_node_id = node->id;
+	}
+
+	/* Query all nodes in the cluster to find the best source node */
+	{
+		Oid			current_best_node_id = InvalidOid;
+		XLogRecPtr	current_best_lsn = InvalidXLogRecPtr;
+		TimestampTz	current_best_commit_ts = 0;
+		int			source_count = 0;
+		int			tie_count = 0;
+		int			ret;
+		int			i;
+
+		/* Connect to SPI to query all nodes */
+		if (SPI_connect() != SPI_OK_CONNECT)
+			elog(ERROR, "SPI_connect failed");
+
+		/* Get all nodes and their interfaces from the spock schema */
+		ret = SPI_execute("SELECT n.node_id, n.node_name, i.if_dsn "
+						  "FROM spock.node n "
+						  "JOIN spock.node_interface i ON i.if_nodeid = n.node_id "
+						  "WHERE i.if_name = 'default'",
+						  true, 0);
+
+		if (ret == SPI_OK_SELECT && SPI_processed > 0)
+		{
+			TupleDesc	tupdesc = SPI_tuptable->tupdesc;
+
+			/* Query each surviving node to see what data it has from the failed origin */
+			for (i = 0; i < SPI_processed; i++)
+			{
+				HeapTuple	tuple = SPI_tuptable->vals[i];
+				Oid			node_id;
+				char	   *node_name;
+				char	   *node_dsn;
+				XLogRecPtr	remote_lsn = InvalidXLogRecPtr;
+				TimestampTz	remote_ts = 0;
+				bool		has_progress = false;
+				bool		isnull;
+
+				/* Get node information */
+				node_id = DatumGetObjectId(SPI_getbinval(tuple, tupdesc, 1, &isnull));
+				if (isnull)
+					continue;
+
+				node_name = SPI_getvalue(tuple, tupdesc, 2);
+				if (!node_name)
+					continue;
+
+				node_dsn = SPI_getvalue(tuple, tupdesc, 3);
+				if (!node_dsn)
+					continue;
```

Comment:
any of these value having NULL means we have some corruption of data or system is not configured and should be considered as failure